### DISCUSSION TITLE
Add the SSL option of disabling DTLS record replay detection

### DISCUSSION THREAD

**Discussion by huiyuexu** [openssl/openssl#22970] [General Discussion]
As described in RFC 6374 [section-3.3](https://datatracker.ietf.org/doc/html/rfc6347#section-3.3), replay detection is optional.

```
3.3.  Replay Detection
   DTLS optionally supports record replay detection.  The technique used
   is the same as in IPsec AH/ESP, by maintaining a bitmap window of
   received records.  Records that are too old to fit in the window and
   records that have previously been received are silently discarded.
   The replay detection feature is optional, since packet duplication is
   not always malicious, but can also occur due to routing errors.
   Applications may conceivably detect duplicate packets and accordingly
   modify their data transmission strategy.
```

In some cases, apps can detect duplicate packets in BIO level based on their strategy, such as larger window range. Could OpenSSL add the SSL option of disabling DTLS record replay detection?

**Comment by mattcaswell** [comment]
The record replay check is implemented here:

https://github.com/openssl/openssl/blob/504ff2a4ef5f26990a48ca3d664ac1e5d9cb20b9/ssl/record/methods/dtls_meth.c#L40-L59

It would be technically feasible to add an option to turn this off. But I think this is unlikely to be a priority for the dev team.

**Reply by huiyuexu** [reply]
I plan to add the SSL option of disabling DTLS record replay detection as `SSL_OP_NO_DTLS_RECORD_REPLAY_CHECK` and add description in `doc/man3/SSL_CTX_set_options.pod` as

```
=item SSL_OP_NO_DTLS_RECORD_REPLAY_CHECK

Turn off DTLS replay detect as described in RFC 6374 section 4.1.2.6.
By default, replay detect is built in DTLS record handling. Some 
applications may be able to mitigate the replay risks in other ways and 
in such cases the detection is not required. Those applications can turn
this feature off by setting this option. Only affects DTLS over UDP connections.
```
Do you have any suggestions?

**Comment by huiyuexu** [comment]
PR: #22980 @mattcaswell
